### PR TITLE
Add basic support for resource quota

### PR DESCRIPTION
**Pull request by atantawi**
Handles phase 1 of issue #44, namely Basic handling of quota resources {cpu, memory, ...}, {requests.cpu, requests.memory, ...} and {limits.cpu, limits.memory, ...}, including extended resources.

### CODE REVIEW

File: deployment/mcad-controller/templates/clusterrole.yaml [hunk 1]
```diff
@@ -218,3 +218,11 @@ rules:
   - subjectaccessreviews
   verbs:
   - create
+- apiGroups:
+  - ""
```
Comment (dgrove-oss) [inline deployment/mcad-controller/templates/clusterrole.yaml:222]:
```suggestion
  -  workload.codeflare.dev
```

Comment (dgrove-oss) [reply deployment/mcad-controller/templates/clusterrole.yaml:222]: I think it should match the change to config/rbac/role.yaml,

Comment (dgrove-oss) [reply deployment/mcad-controller/templates/clusterrole.yaml:222]: I got it backwards; will update

File: config/rbac/role.yaml [hunk 2]
```diff
@@ -220,3 +220,17 @@ rules:
   - get
   - patch
   - update
+- apiGroups:
+  - workload.codeflare.dev
```
Comment (dgrove-oss) [inline config/rbac/role.yaml:224]:
```suggestion
  - ""
```

File: internal/controller/appwrapper_controller.go [hunk 3]
```diff
@@ -34,6 +34,8 @@ import (
 //+kubebuilder:rbac:groups=workload.codeflare.dev,resources=appwrappers,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=workload.codeflare.dev,resources=appwrappers/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=workload.codeflare.dev,resources=appwrappers/finalizers,verbs=update
+//+kubebuilder:rbac:groups=workload.codeflare.dev,resources=resourcequotas,verbs=get;list;watch
```
Comment (dgrove-oss) [inline internal/controller/appwrapper_controller.go:37]:
```suggestion
//+kubebuilder:rbac:groups="",resources=resourcequotas,verbs=get;list;watch
```

File: internal/controller/appwrapper_controller.go [hunk 4]
```diff
@@ -34,6 +34,8 @@ import (
...
+//+kubebuilder:rbac:groups=workload.codeflare.dev,resources=resourcequotas/status,verbs=get
```
Comment (dgrove-oss) [inline internal/controller/appwrapper_controller.go:38]:
```suggestion
//+kubebuilder:rbac:groups="",resources=resourcequotas/status,verbs=get
```

**Comment by atantawi** [comment]
Fixed kubebuilder directive in `internal/controller/appwrapper_controller.go` so that `config/rbac/role.yaml` and `deployment/mcad-controller/templates/rbac/clusterrole.yaml` are matched.
(Also, rebased on main to pick up commit #141.)

Now E2E test fails due to:
`Internal error occurred: failed calling webhook "webhook.cert-manager.io": failed to call webhook: Post "https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=10s": dial tcp 10.96.50.154:443: connect: connection refused`

**Comment by atantawi** [comment]
> Fixed kubebuilder directive in `internal/controller/appwrapper_controller.go` so that `config/rbac/role.yaml` and `deployment/mcad-controller/templates/rbac/clusterrole.yaml` are matched. (Also, rebased on main to pick up commit #141.)
> 
> Now E2E test fails due to: `Internal error occurred: failed calling webhook "webhook.cert-manager.io": failed to call webhook: Post "https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=10s": dial tcp 10.96.50.154:443: connect: connection refused`

Fixed with commit #144

File: config/rbac/role.yaml (hunk 2 above)
Comment (atantawi) [reply config/rbac/role.yaml]: Done

File: internal/controller/appwrapper_controller.go (hunk 3 above)
Comment (atantawi) [reply internal/controller/appwrapper_controller.go]: Done

File: internal/controller/appwrapper_controller.go (hunk 4 above)
Comment (atantawi) [reply internal/controller/appwrapper_controller.go]: Done

File: internal/controller/dispatcher.go [hunk 5]
```diff
@@ -257,3 +265,41 @@ func (r *Dispatcher) dispatch(ctx context.Context) (ctrl.Result, error) {
...
+func (r *Dispatcher) getUnadmittedAppWrappersWeights(ctx context.Context) (map[string]*WeightsPair, error) {
+	appWrappers := &mcadv1beta1.AppWrapperList{}
+	if err := r.List(ctx, appWrappers, client.UnsafeDisableDeepCopy); err != nil {
+		return nil, err
+	}
+	weightsPairMap := make(map[string]*WeightsPair)
+	for _, appWrapper := range appWrappers.Items {
+		phase, step := r.getCachedAW(&appWrapper)
+		if phase == mcadv1beta1.Running &&
+			step == mcadv1beta1.Idle || step == mcadv1beta1.Creating || step == mcadv1beta1.Created {
```
Comment (tardieu) [inline internal/controller/dispatcher.go:280]: This condition does not seem right. Just use: `step != mcadv1beta1.Idle`.

File: internal/controller/dispatch_logic.go [hunk 6]
```diff
@@ -237,6 +262,18 @@ func aggregateRequests(appWrapper *mcadv1beta1.AppWrapper) Weights {
 	return request
 }
 
+// Aggregate limits
+func aggregateLimits(appWrapper *mcadv1beta1.AppWrapper) Weights {
+	limit := Weights{}
+	for _, r := range appWrapper.Spec.Resources.GenericItems {
+		for _, cpr := range r.CustomPodResources {
+			// TODO: check why not implemented limits spec in CustomPodResource?
+			limit.AddProd(cpr.Replicas, NewWeights(cpr.NotImplemented_Limits))
```
Comment (tardieu) [inline internal/controller/dispatch_logic.go:271]: Now that limits are relevant, please update the CRD renaming `NotImplemented_Limit` to `Limits`.

File: internal/controller/quota_tracker.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	if unAdmittedWeights, exists = tracker.unAdmittedWeightsMap[namespace]; exists {
+		quotaWeights.Sub(unAdmittedWeights)
+	}
+	quotaFits, insufficientResources := appWrapperAskWeights.Fits(quotaWeights)
+
+	mcadLog.Info("QuotaTracker.Satisfies():", "namespace", namespace,
+		"QuotaWeights", quotaState.quota, "UsedWeights", quotaState.used,
+		"AllocatedWeights", quotaState.allocated, "unAdmittedWeights", unAdmittedWeights,
+		"AvailableWeights", quotaWeights, "appWrapperAskWeights", appWrapperAskWeights,
+		"quotaFits", quotaFits)
```
Comment (tardieu) [inline internal/controller/quota_tracker.go:97]: Please disable this log line by default as it is going to make logging very verbose.

File: internal/controller/dispatcher.go [hunk 8]
```diff
@@ -257,3 +265,41 @@ func (r *Dispatcher) dispatch(ctx context.Context) (ctrl.Result, error) {
 
 	return ctrl.Result{RequeueAfter: dispatchDelay}, nil
 }
+
+// Calculate resource demands of appWrappers that have been dispatched but haven't
+// passed through ResourceQuota admission controller yet (approximated by resources not created yet)
+func (r *Dispatcher) getUnadmittedAppWrappersWeights(ctx context.Context) (map[string]*WeightsPair, error) {
```
Comment (tardieu) [inline internal/controller/dispatcher.go:271]: Maybe rename to `getUnadmittedAppWrapperPodsWeights`?

File: internal/controller/dispatch_logic.go [hunk 9]
```diff
@@ -204,13 +204,38 @@ func (r *Dispatcher) selectForDispatch(ctx context.Context) ([]*mcadv1beta1.AppW
...
+		if err := r.List(ctx, resourceQuotas, client.UnsafeDisableDeepCopy,
+			&client.ListOptions{Namespace: namespace}); err != nil {
+			return nil, err
+		}
+		quotaFits := true
+		var appWrapperAskWeights *WeightsPair
+		insufficientResources := []v1.ResourceName{}
+		if len(resourceQuotas.Items) > 0 {
+			appWrapperAskWeights = getWeightsPairForAppWrapper(appWrapper)
+			// TODO: relax assumption that only one resourceQuota per nameSpace
```
Comment (tardieu) [inline internal/controller/dispatch_logic.go:219]: Please open an issue (or comment on an existing issue) to document this limitation and others.

**Review by tardieu — CHANGES_REQUESTED** [review]

File: internal/controller/dispatcher.go (hunk 5 above)
Comment (atantawi) [reply internal/controller/dispatcher.go]: done

File: internal/controller/dispatch_logic.go (hunk 6 above)
Comment (atantawi) [reply internal/controller/dispatch_logic.go]: done

File: internal/controller/quota_tracker.go (hunk 7 above)
Comment (atantawi) [reply internal/controller/quota_tracker.go]: done

File: internal/controller/dispatcher.go (hunk 8 above)
Comment (atantawi) [reply internal/controller/dispatcher.go]: done

File: internal/controller/dispatch_logic.go (hunk 9 above)
Comment (atantawi) [reply internal/controller/dispatch_logic.go]: added to issue #44 .

**Review by tardieu — APPROVED** [review]
lgtm

**Pull request merged**